8275644: Replace VMReg in shuffling code with something more fine grained.

JornVernee opened the pull request:
This patch replaces `VMReg` with a new `VMStorage` class in the C++ code, which can be used to represent stack arguments of different sizes, as well as allowing finer grained stack addressing.

The old masm code that used `VMReg` for moving values is also replaced with new code that uses `VMStorage`. And `VMReg` is changed to `VMStorage` everywhere.

The `vmstorageBase` class was added in C++ to mirror the `jdk.internal.foreign.VMStorage`. That class now also has an additional field to encode the size of a stack argument, or a register mask, which can also be used to encode the size of a register argument. This also leads to some stubs needing to be updated on unsupported platforms.

Lastly, instead of encoding stack indexes as 64-bit slots as we did before, they are now encoded as byte offsets, which allows finer grained stack addressing. (this leads to a bunch of values being multiplied by 8 in the Java code as well).

---

This change is needed to address https://bugs.openjdk.org/browse/JDK-8275584 as well as adding support for `long double` as well as more exotic ABIs that take arguments as vectors.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8275644](https://bugs.openjdk.org/browse/JDK-8275644): Replace VMReg in shuffling code with something more fine grained.


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer) ⚠️ Review applies to [513ccc02](https://git.openjdk.org/panama-foreign/pull/699/files/513ccc02ed15e3b51d73458514e7f33d4d376445)
 * [Nick Gasson](https://openjdk.org/census#ngasson) (@nick-arm - Committer) ⚠️ Review applies to [7d54e3d5](https://git.openjdk.org/panama-foreign/pull/699/files/7d54e3d59b3521ff69f8a107f7aba887fe48890e)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/699/head:pull/699` \
`$ git checkout pull/699`

Update a local copy of the PR: \
`$ git checkout pull/699` \
`$ git pull https://git.openjdk.org/panama-foreign pull/699/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 699`

View PR using the GUI difftool: \
`$ git pr show -t 699`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/699.diff">https://git.openjdk.org/panama-foreign/pull/699.diff</a>

</details>

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/VMStorage.java:
```diff
@@ -27,21 +27,35 @@
 public class VMStorage {
-    private final int type;
+    private final byte type;
+    private final short segmentMaskOrSize;
```
We have a couple of fields here which do double duty - e.g. depending on whether the storage models a stack argument or a record, we end up with different meanings. One way is to use (as you did) names like `segmentMaskOrSize`.

But perhaps a better (?) way would be to make VMStorage an abstract class/interface and then have two implementations, for stack slots and registers. Although I realize this might impact the VM code in a bad way (as I think this class is mirrored there in some way).

mcimadamore replied:
That's ok, I suspected something similar was at play. In that case I suggest a uniform naming scheme, plus perhaps some documents explaining the semantics of the fields in the various cases.

JornVernee replied:
I can add some comments to the fields. What do you mean by 'uniform naming scheme'?

mcimadamore replied:
> I can add some comments to the fields. What do you mean by 'uniform naming scheme'?

Sorry, I meant that one field is called `segmentMaskOrSize` but the other is just called `index`. Using `or` on both might be a good idea.

JornVernee replied:
Ah, ok. That sounds good.

mcimadamore reviewed (approved):
Java code looks good - left some minor comment.
I like the generality that this approach buys, I think it will be helpful when modelling more exotic ABIs/calling conventions.

JornVernee commented:
I'm still testing but I added some comments, and changed the name to indexOrOffset/index_or_offset on both sides. (I also got rid of the union in the C++ code, since it didn't really buy us anything)

JornVernee commented:
@nick-arm Could you take a look as well?

I've also taken a stab at fixing [Incorrect stack spilling in CallArranger on MacOS/AArch64](https://bugs.openjdk.org/browse/JDK-8275584)
Here: https://github.com/JornVernee/panama-foreign/compare/VMStorage2...JornVernee:panama-foreign:Mac_Stack
But there are still some failing tests.

nick-arm commented on src/hotspot/cpu/aarch64/foreignGlobals_aarch64.cpp:
```diff
@@ -42,112 +43,165 @@ bool ABIDescriptor::is_volatile_reg(FloatRegister reg) const {
-      case T_CHAR:
-      case T_INT:
-        masm->move32_64(from_vmreg, to_vmreg, tmp_reg);
+constexpr int RFP_BIAS = 16; // skip old rfp and lr
```
Should this be `static constexpr`?

JornVernee replied:
Can do.

nick-arm commented on src/hotspot/cpu/aarch64/vmstorage_aarch64.inline.hpp:
```diff
@@ -0,0 +1,83 @@
+ */
+
+#ifndef SHARE_PRIMS_VMSTORAGE_AARCH64
+#define SHARE_PRIMS_VMSTORAGE_AARCH64
```
Should this be `CPU_AARCH64_VMSTORAGE_AARCH64_HPP`?

JornVernee replied:
Yes. I messed up this pattern for all the new header files. Thanks for catching!

nick-arm reviewed (approved):
The AArch64-specific changes look sensible to me!

JornVernee commented:
/integrate
